### ISSUE TITLE
New Builder Template containing animated gifs

### ISSUE THREAD

**Issue by g-no** [isolver/OpenHandWrite#194]
A new PsychoPy Builder Example / Template would be great. Actually these would be two templates: 1 for using a TabletPC (writing on the screen as e.g. MS Surface), 2. for using a monitor and an external tablet (writing on paper or tablet surface, e.g. Wacom Intuos).

_legend for showpen:

- P = Pen tip visible
- C = Collect data
- T = show pen Trace
- A = show Angle/tilt
- G = show Glow around pen tip (indicating pressure) [A and G: priority 2]_

_no pen trace visible on buttons_

The templates should contain the following (description for TabletPC, maybe no changes are needed for an external tablet?):

+ validation
+ PreTrial 1
++ Step 1: blank screen, background color: grey, showpen: PC, show picture "NoWriting" in all four corners
++ Step 2: background color: grey, showpen: PC, show picture "NoWriting" in all four corners, after delay 2 sec play audio description ("try using this pen bla bla")
++ Step 3: when audio description has ended, change background color to white, remove pictures "NoWriting" (show blank screen), show button NEXT (image in bottom/right corner), showpen PTCAG, wait for button NEXT, then goto next Trial
+ PreTrial 2
++ Step 1: blank screen, background color: grey, showpen: PC, show picture "NoWriting" in all four corners
++ Step 2: background color: grey, showpen: PC, show picture "NoWriting" in all four corners, show image "Lineatur1", after delay 1 sec play audio description ("write your name on the line")
++ Step 3: show image "Lineatur1", when audio description has ended, change background color to white, remove pictures "NoWriting", show button NEXT (image in bottom/right corner), showpen PTCAG, wait for button NEXT, then goto next Trial
+ Trial 1
++ Step 1: blank screen, background color: grey, showpen: PC, show picture "NoWriting" in all four corners
++ Step 2: background color: grey, showpen: PC, show picture "NoWriting" in all four corners, show image "Lineatur3", after delay 1 sec play audio description ("write all letters of the ABC from A to Z in upper case")
++ Step 3: show image "Lineatur3", when audio description has ended, change background color to white, remove pictures "NoWriting", show button NEXT (image in bottom/right corner), showpen PTCAG, wait for button NEXT, then goto next Trial
+ Trial 2
++ Step 1: blank screen, background color: grey, showpen: PC, show picture "NoWriting" in all four corners
++ Step 2: background color: grey, showpen: PC, show picture "NoWriting" in all four corners, show image "Lineatur3", after delay 1 sec play audio description ("when I say 'go' write the letters of the ABC in upper case as fast as you can for 15 second until I say 'stop'")
++ Step 3: show image "Lineatur3", when audio description has ended, delay 1 second, show image "3" centered, delay 1 second, show image "2", delay 1 second, show image "1", delay 1 second, change background color to white, remove pictures "NoWriting", play sound "go", showpen PTCAG, delay 15 seconds then showpen PC, show button NEXT (image in bottom/right corner), wait for button NEXT, then goto next Trial
+ Trial 3
++ Step 1: blank screen, background color: grey, showpen: PC, show picture "NoWriting" in all four corners
++ Step 2: background color: grey, showpen: PC, show picture "NoWriting" in all four corners, show image "Circles CCW anim" (animated gif), after delay 1 sec play audio description ("please draw about 10 to 20 fast circles around the green dot")
++ Step 3: when both audio description and animation have ended, delay 1 second, show image "green dot", delay 1 second, change background color to white, remove pictures "NoWriting", showpen PTCAG, show button NEXT (image in bottom/right corner), wait for button NEXT, then goto next Trial
+ Trial 4 to 10
same as trial 3 but with different images and audio files
+ PostTrial
++ Step 1: blank screen, background color: grey, showpen: PC, delay 1 second, show image "thankyou", show button END, wait for button END, then close program

I'll update this post with an animated gif.

**Comment by isolver** [comment]
- Can other movie file formats be used instead (like .avi, etc)
- Is only the video from the movie file needed; i.e. can any audio in the video fine be ignored, or should it be played too if present.
- to confirm: you want to be able to draw the pen position / graphics on top of the video frames?

**Comment by isolver** [comment]
Does this need to be a Builder example, or can we extend the Coder example instead if it turns out to be much quicker to implement in Coder after I've fully digested all this? ;)

For example, things "like no pen trace visible on buttons" will likely require messy custom code in Builder and would be much simpler to do directly in a python experiment script.

**Comment by g-no** [comment]
Here's a mp4 video that's running with 50 Hz. There's no need for animated gif if mp4 works. [I just found out that mp4 files are not supported. I'll zip it.]

The video doesn't contain any audio.

No need to draw on top of the video.

The "no pen trace visible on buttons" is not important.

It would be much nicer to have a Builder example, because then other people (including me) would be able to alter the template to their needs more easily. So if any of the above requirements can't be done easily within Builder, we can discuss whether to dismiss them or drop the Builder account.

[Circles CCW.zip](https://github.com/isolver/OpenHandWrite/files/1686178/Circles.CCW.zip)

**Comment by isolver** [comment]
Sounds good.

**Comment by isolver** [comment]
@g-no:

1. what do you mean by the 'validation' step; is this something I wrote for you before? ;)

2. 'PreTrial 1' and 'PreTrial 2': should only one of  these is to occur per experiment trial?  Should which be controlled by a trial level variable or will all trials in an experiment only use v1 or v2?

3. 'Trial 1', 'Trial 2', 'Trial 3':  should only one of  these is to occur per experiment trial?  Should which be controlled by a trial level variable or will all trials in a specific experiment only use one of the 3 trial types?

4. "show picture "NoWriting" all four corners": So display the same picture 4 times on the screen; each filling 1/4 of the screen? Should it be an actual image, or just test written on the screen in 4 location?

**Comment by g-no** [comment]
> what do you mean by the 'validation' step; is this something I wrote for you before? ;)

I meant the validation process that's built into the Experiment Template. It's in getwrite\experiments\Experiment Template\run.py starting with line 636.

> 'PreTrial 1' and 'PreTrial 2': should only one of these is to occur per experiment trial? Should which be controlled by a trial level variable or will all trials in an experiment only use v1 or v2? 'Trial 1', 'Trial 2', 'Trial 3': should only one of these is to occur per experiment trial? Should which be controlled by a trial level variable or will all trials in a specific experiment only use one of the 3 trial types?

It doesn't really matter if this is called PreTrial or just Trial. In this template all of these trial types should occur. So far, only the trial3 type will be used for several trials, but in general all types should give the opportunity to be recycled.

> "show picture "NoWriting" all four corners": So display the same picture 4 times on the screen; each filling 1/4 of the screen? Should it be an actual image, or just test written on the screen in 4 location?

It's actually an image (see below). It was just a spontaneous idea to place a small (ca. 1/10 of screen height) copy in each of the corners in order to show something else in the center of the screen.
![nowriting](https://user-images.githubusercontent.com/15689032/35769140-cf4f3fc6-0906-11e8-8897-a68bca7384b2.jpg)

**Comment by isolver** [comment]
Is PreTrial 1 the same as PreTrial 2 except no center image is shown? If so, then PreTrial 1 and PreTrial 2 can really just be the same, where PreTrial 1 image is set to blank. Is this correct?

Thanks

**Comment by g-no** [comment]
Yes, that's correct.

**Comment by isolver** [comment]
Are PreTrial1 and PreTrial2 examples of 'practice' trials that you want to occur prior to any of the trial 1, 2, etc logic? Before I thought they represented logic you wanted to run at the start of each trial prior to the main trial logic; this is what was confusing me. 

@g-no: Can you confirm the intent is to show some practice trials prior to the block of main data collection trials?

**Comment by isolver** [comment]
BTW, I have a Builder example that plays a .mp4 video and a .wav file during each trial. It is just a test, not the trial paradigm you list above, but at least we know we can play them via Builder. 

For the mp4, I did need to patch the version of PsychoPy we are using with a fix for a movie3.py  issue.

**Comment by g-no** [comment]
Yes, the intent is to show some practice trials prior to the block of main data collection trials.

**Comment by g-no** [comment]
Thinking about the experiment, it would be good to have a "repeat trial" option built in somehow. If, e.g. a participant accidentally moves on to the next trial, it would be good to have the option to go back. Sorry, this is not more than thinking aloud ...
